soroban-env-host/src/auth.rs

```diff
@@ -0,0 +1,946 @@
+            if let Ok(invoker_contract) = host.get_invoking_contract_internal() {
+                if &invoker_contract == contract_addr {
+                    return Ok(());
+                }
+            }
+        }
+
+        if let Some(curr_invocation) = self.call_stack.last() {
+            match &mut self.mode {
+                AuthorizationMode::Enforcing => {
+                    // Iterate all the trackers and try to find one that
+                    // fullfills the authorization requirement.
+                    for tracker in &mut self.trackers {
+                        let address_matches = if let Some(addr) = &tracker.address {
+                            addr == &address
+                        } else {
+                            // Lazily fill the address for the invoker trackers,
+                            // so that it's possible to create the auth manager
+                            // without knowing the invoker beforehand and also
+                            // to not keep calling into `source_account` function.
+                            let source_addr = ScAddress::Account(host.source_account()?);
+                            let source_matches = source_addr == address;
+                            tracker.address = Some(source_addr);
+                            source_matches
+                        };
+                        // If address doesn't much, just skip the tracker.
+                        if address_matches {
+                            match tracker.maybe_authorize_invocation(
+                                host,
+                                &curr_invocation.contract_id,
+                                curr_invocation.function_name,
+                                &args,
+                            ) {
+                                // If tracker doesn't have a matching invocation,
+                                // just skip it (there could still be another
+                                // tracker  that matches it).
+                                Ok(false) => continue,
+                                // Found a matching authorization.
+                                Ok(true) => return Ok(()),
+                                // Found a matching authorization, but another
+                                // requirement hasn't been fullfilled (for
+                                // example, incorrect authentication or nonce).
+                                Err(e) => return Err(e),
+                            }
+                        }
+                    }
+                    // No matching tracker found, hence the invocation isn't
+                    // authorized.
+                    Err(host.err_general("invocation is not authorized"))
+                }
+                AuthorizationMode::Recording(recording_info) => {
+                    if let Some(tracker_id) = recording_info
+                        .tracker_by_address_handle
```

Comment:
Nice catch, I've actually noticed this only when ran some tests. This should be fixed now (as well as another recording-related bug)

Reply:
To be clear, the idea is to completely reset the whole `AuthorizationManager`. The only reason for not creating a new instance is to store the previous authorizations in order to be able to verify them. Maybe it would actually make more sense to do the backup one level higher (i.e. in the host). Does that sound reasonable?

Reply:
Is `testutils` required for recording mode? I see you added `tracker_by_address_handle.clear()` in `AuthorizationManager::reset()`, but that's behind a `testutils` conditional.

And I don't think I understand your question (I think my misunderstanding here is related to the above point). The host already calls `AuthorizationManager::reset()` right?

Reply:
> Is `testutils` required for recording mode?

Recording mode will be used by non-tests and needs to exist outside the testutils feature. For example the soroban-rpc imports the soroban-env-host and will need to use the recording mode.

Reply:
> Is testutils required for recording mode? I see you added tracker_by_address_handle.clear() in AuthorizationManager::reset(), but that's behind a testutils conditional.

The whole `reset` method is test-specific, hence it's only enabled when `testutils` feature is present. The common uint testing pattern is to reuse the same host instance for multiple contract invocations, which is why we need to reset auth for every invocation in order to emulate the real behavior. In not-test mode both enforcing and recording auth shouldn't be reset - instead different instance of host would be needed. This may or may not need an adjustment when we allow multi-op transactions, but for now this reflects the current state of the things.

Reply:
So I think I got a bit off track here - you mentioned that you fixed the bug at the top of this thread by resetting `AuthorizationManager`. Where is that fix exactly?

Reply:
The fix is in this commit: https://github.com/stellar/rs-soroban-env/pull/645/commits/4a4a8d43b1e7fa941e932041d6a42f6f28b466cf

What I'm thinking about is to remove the `reset` method completely and instead keep a backup of `AuthorizationManager` in tests.

Reply:
So you're saying this bug can only happen in tests?

Reply:
Yeah, in non-test environment there is nothing to reset in the first place.

Reply:
As we discussed offline, `tracker_by_address_handle` needs to be part of the auth managers snapshot.